gh-103921: Document PEP 695 (#104642)
python/cpython: Doc/reference/executionmodel.rst

that unbound local variables are looked up in the global namespace.
The namespace of the class definition becomes the attribute dictionary of
the class. The scope of names defined in a class block is limited to the
class block; it does not extend to the code blocks of methods. This includes
comprehensions and generator expressions, but it does not include
:ref:`annotation scopes <annotation-scopes>`,
which have access to their enclosing class scopes.
This means that the following will fail::

   class A:
       a = 42
       b = list(a + i for i in range(10))

However, the following will succeed::

   class A:
       type Alias = Nested
       class Nested: ...

   print(A.Alias.__value__)  # <type 'A.Nested'>

.. _annotation-scopes:

Annotation scopes
-----------------

:ref:`Type parameter lists <type-params>` and :keyword:`type` statements
introduce *annotation scopes*, which behave mostly like function scopes,
but with some exceptions discussed below. :term:`Annotations <annotation>`

> currently do not use annotation scopes, but they are expected to use
> annotation scopes in Python 3.13 when :pep:`649` is implemented.

Reviewer: Not sure we should be documenting things we _expect_ to happen, because they might not happen. We can add a note about how annotations use annotation scopes when PEP-649 is implemented.

```suggestion
currently do not use annotation scopes.
```

Author: I would like to leave some indication of why we use the term "annotation scopes" when annotations don't use annotation scopes.

Reviewer: OK

Annotation scopes are used in the following contexts:

* Type parameter lists for :ref:`generic type aliases <generic-type-aliases>`.

> * Type parameter lists for :ref:`generic functions <generic-functions>`. The function's annotations are

Reviewer: ```suggestion
* Type parameter lists for :ref:`generic functions <generic-functions>`.
  A generic function's annotations are
```

  executed within the annotation scope, but its defaults and decorators are not.

> * Type parameter lists for :ref:`generic classes <generic-classes>`. The class's base classes and

Reviewer: ```suggestion
* Type parameter lists for :ref:`generic classes <generic-classes>`.
  A generic class's base classes and
```

  keyword arguments are executed within the annotation scope, but its decorators are not.
* The bounds and constraints for type variables
  (:ref:`lazily evaluated <lazy-evaluation>`).
* The value of type aliases (:ref:`lazily evaluated <lazy-evaluation>`).

Annotation scopes differ from function scopes in the following ways:

* If an annotation scope is immediately within a class scope, or within another
  annotation scope that is immediately within a class scope, names defined in the
  class scope can be accessed from within the annotation scope, as if the code in the
  annotation scope was executing directly within the class scope. (By contrast, regular

>   functions defined within classes cannot access names defined in the class scope.)

Reviewer: Shorter sentences:

```suggestion
  class scope can be accessed from within the annotation scope. The code in the
  annotation scope is executed in the same way as if it were executed directly
  within the class scope. This contrasts with regular
  functions defined within classes, which cannot access names defined in the class scope.
```

Author: Pushing a slightly different rewrite; I feel "executed in the same way as if it were executed directly within the class scope" isn't quite true.

* Expressions in annotation scopes cannot contain :keyword:`yield`, ``yield from``,

>   :keyword:`await`, or :token:`:= <~python-grammar:expression>` expressions. (These

Reviewer: Is this the link you wanted? I think you might have wanted:

```suggestion
  :keyword:`await`, or :token:`:= <~python-grammar:assignment_expression>`
  expressions. (These
```

I'm also not sure the tilde is necessary here? I think this might work equally well. But not 100% confident about that:

```suggestion
  :keyword:`await`, or :token:`:= <python-grammar:assignment_expression>`
  expressions. (These
```

Author: Your second suggestion works, thanks.

Author: (And probably the first one too but I didn't try.)

  expressions are allowed in other scopes contained within the annotation scope.)
* Names defined in annotation scopes cannot be rebound with :keyword:`nonlocal`
  statements in inner scopes. This includes only type parameters, as no other
  syntactic elements that can appear within annotation scopes can introduce new names.

> * While annotation scopes internally have a name, that name is not reflected in the

Reviewer: ```suggestion
* While annotation scopes have an internal name, that name is not reflected in the
```

>   ``__qualname__`` of objects defined within the scope. Instead, the ``__qualname__``

Reviewer: Link to https://docs.python.org/3/glossary.html#term-qualified-name

```suggestion
  :term:`__qualname__ <qualified name>` of objects defined within the scope.
  Instead, the ``__qualname__``
```

>   of such objects is as if the object was defined in the enclosing scope.

Reviewer: ```suggestion
  of such objects is as if the object were defined in the enclosing scope.
```

Author: Keeping the subjunctive alive

.. versionadded:: 3.12
   Annotation scopes were introduced in Python 3.12 as part of :pep:`695`.

.. _lazy-evaluation:

Lazy evaluation
---------------

> The values of type aliases created through the :keyword:`type` statement and
> the bounds and constraints of type variables created through the
> :ref:`type parameter syntax <type-params>` are *lazily evaluated*. This means
> that they are not evaluated when the type alias or type variable is created,
> but only when evaluation is necessary to resolve an attribute access.

Reviewer: Shorter sentences:

```suggestion
The values of type aliases created through the :keyword:`type` statement are
*lazily evaluated*. The same applies to the bounds and constraints of type
variables created through the :ref:`type parameter syntax <type-params>`.
This means that they are not evaluated when the type alias or type variable is
created. Instead, they are only evaluated when doing so is necessary to resolve
an attribute access.
```

Example:

.. doctest::

   >>> type Alias = 1/0
   >>> Alias.__value__
   Traceback (most recent call last):
     ...
   ZeroDivisionError: division by zero

> Here the exception is raised only when the ``__value__`` attribute is accessed.

Reviewer: Could we also have an example using `TypeVar` bounds?

Author: Adding

that unbound local variables are looked up in the global namespace.
The namespace of the class definition becomes the attribute dictionary of
the class. The scope of names defined in a class block is limited to the
class block; it does not extend to the code blocks of methods. This includes
comprehensions and generator expressions, but it does not include
:ref:`annotation scopes <annotation-scopes>`,
which have access to their enclosing class scopes.
This means that the following will fail::

   class A:
       a = 42
       b = list(a + i for i in range(10))

However, the following will succeed::

   class A:
       type Alias = Nested

>        class Nested: ...

Reviewer: ```suggestion
       class Nested: pass
```

.. _annotation-scopes:

Annotation scopes
-----------------

:ref:`Type parameter lists <type-params>` and :keyword:`type` statements
introduce *annotation scopes*, which behave mostly like function scopes,
but with some exceptions discussed below. :term:`Annotations <annotation>`
currently do not use annotation scopes, but they are expected to use
annotation scopes in Python 3.13 when :pep:`649` is implemented.

Annotation scopes are used in the following contexts:

* Type parameter lists for :ref:`generic type aliases <generic-type-aliases>`.
* Type parameter lists for :ref:`generic functions <generic-functions>`.
  A generic function's annotations are
  executed within the annotation scope, but its defaults and decorators are not.
* Type parameter lists for :ref:`generic classes <generic-classes>`.
  A generic class's base classes and
  keyword arguments are executed within the annotation scope, but its decorators are not.
* The bounds and constraints for type variables
  (:ref:`lazily evaluated <lazy-evaluation>`).
* The value of type aliases (:ref:`lazily evaluated <lazy-evaluation>`).

Annotation scopes differ from function scopes in the following ways:

> * If an annotation scope is immediately within a class scope, or within another
>   class scope, the code in the

Reviewer: What's meant by "within another class scope"? Did you mean something like:

```suggestion
* If an annotation scope is immediately within a class scope, or within a
  class scope within another class scope, the code in the
```

Author: It's about annotation scopes within annotation scopes within class scopes. Concretely, that happens with TypeVar bounds. I'll clarify that and add a short introductory sentence as this one is very complex.

Reviewer: Thanks, the new wording looks great!

>   class scope can be accessed from within the annotation scope. The code in the

Reviewer: The first sentence here no longer makes grammatical sense after your latest push :(

Author: Thanks, fixed

  annotation scope can use names defined in the class scope as if it were
  executed directly within the class body. This contrasts with regular
  functions defined within classes, which cannot access names defined in the class scope.
* Expressions in annotation scopes cannot contain :keyword:`yield`, ``yield from``,
  :keyword:`await`, or :token:`:= <python-grammar:assignment_expression>`
  expressions. (These expressions are allowed in other scopes contained within the
  annotation scope.)
* Names defined in annotation scopes cannot be rebound with :keyword:`nonlocal`
  statements in inner scopes. This includes only type parameters, as no other
  syntactic elements that can appear within annotation scopes can introduce new names.
* While annotation scopes have an internal name, that name is not reflected in the
  :term:`__qualname__ <qualified name>` of objects defined within the scope.

>   Instead, the ``__qualname__``

Reviewer: ```suggestion
  Instead, the :attr:`!__qualname__`
```

  of such objects is as if the object were defined in the enclosing scope.

.. versionadded:: 3.12
   Annotation scopes were introduced in Python 3.12 as part of :pep:`695`.

.. _lazy-evaluation:

Lazy evaluation
---------------

The values of type aliases created through the :keyword:`type` statement are
*lazily evaluated*. The same applies to the bounds and constraints of type
variables created through the :ref:`type parameter syntax <type-params>`.
This means that they are not evaluated when the type alias or type variable is
created. Instead, they are only evaluated when doing so is necessary to resolve
an attribute access.

Example:

.. doctest::

   >>> type Alias = 1/0
   >>> Alias.__value__
   Traceback (most recent call last):
     ...
   ZeroDivisionError: division by zero
   >>> def func[T: 1/0](): pass

>    >>> T, = func.__type_params__

Reviewer: The tuple unpacking might be a little subtle here; indexing might make it more obvious

```suggestion
   >>> T = func.__type_params__[0]
```

that unbound local variables are looked up in the global namespace.
The namespace of the class definition becomes the attribute dictionary of
the class. The scope of names defined in a class block is limited to the
class block; it does not extend to the code blocks of methods. This includes
comprehensions and generator expressions since they are implemented using a

> function scope, but it does not include :ref:`annotation scopes <annotation-scopes>`,

Reviewer 2: Let's leave out the part about "implemented using a function scope," since it's no longer true for comprehensions, and isn't critical to the point of this section anyway?

Annotation scopes are used in the following contexts:

* Type parameter lists for generic type aliases.
* Type parameter lists for generic functions. The function's annotations are
  executed within the type scope, but its defaults and decorators are not.
* Type parameter lists for generic classes. The class's base classes and
  keyword arguments are executed within the type scope, but its decorators are not.
* The bounds and constraints for type variables.
* The value of type aliases.

Annotation scopes differ from function scopes in the following ways:

* If an annotation scope is immediately within a class scope, or within another
  annotation scope that is immediately within a class scope, names defined in the
  class scope can be accessed from within the annotation scope, as if the code in the
  annotation scope was executing directly within the class scope. (By contrast, regular
  functions defined within classes cannot access names defined in the class scope.)
* Expressions in annotation scopes cannot contain :keyword:`yield`, ``yield from``,
  :keyword:`await`, or :token:`:= <~python-grammar:expression>` expressions. (These

>   expressions are allowed in other scopes contained within the type scope.)

Reviewer 2: Should we avoid introducing "type scope" as a synonym for "annotation scope"?
```suggestion
  expressions are allowed in other scopes contained within the annotation scope.)
```

Author: I was going to use the term "type scope" at first but didn't catch all places where I had used the term, thanks for catching that. (There are a few more in this file.)

* Names defined in annotation scopes cannot be rebound with :keyword:`nonlocal`
  statements in inner scopes. This includes only type parameters, as no other

>   syntactic elements that can appear within type scopes can introduce new names.

Reviewer 2: ```suggestion
  syntactic elements that can appear within annotation scopes can introduce new names.
```
